src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -1670,14 +1673,23 @@ internal static BoundBlock BindMethodBody(MethodSymbol method, TypeCompilationSt
                     BoundNode methodBodyForSemanticModel = methodBody;
                     NullableWalker.SnapshotManager snapshotManager = null;
                     ImmutableDictionary<Symbol, Symbol> remappedSymbols = null;
-                    if (bodyBinder.Compilation.NullableSemanticAnalysisEnabled)
+                    var compilation = bodyBinder.Compilation;
+                    var isSufficientLangVersion = compilation.LanguageVersion >= MessageID.IDS_FeatureNullableReferenceTypes.RequiredVersion();
+                    switch (isSufficientLangVersion, compilation)
                     {
-                        // Currently, we're passing an empty DiagnosticBag here because the flow analysis pass later will
-                        // also run the nullable walker, and issue duplicate warnings. We should try to only run the pass
-                        // once.
-                        // https://github.com/dotnet/roslyn/issues/35041
-                        methodBodyForSemanticModel = NullableWalker.AnalyzeAndRewrite(bodyBinder.Compilation, method, methodBody, bodyBinder, new DiagnosticBag(), createSnapshots: true, out snapshotManager, ref remappedSymbols);
+                        case (true, { NullableSemanticAnalysisEnabled: true }):
```

Comment:
We used to ignore the language version when running nullable semantic analysis. It didn't affect whether we gave diagnostics, but now it does. An alternative to this solution would be to do the analysis, passing a dummy diagnostic bag for `case (false, { NullableSemanticAnalysisEnabled: true })`.

We can look into potentially giving the nullability diagnostics even if the language version is insufficient, *if* we ShouldRunNullableWalker for example. But I did not want to take on the test churn for it in this PR.

Reply:
> We used to ignore the language version when running nullable semantic analysis. It didn't affect whether we gave diagnostics, but now it does. 

Can we condition those diagnostics to a warning wave instead?

Reply:
To be clear, the "new" diagnostics would only be for an error scenario where a project has a pre-C# 8 language version, but also has NRT enabled using project-level settings or `#nullable` directives in source. It feels like we don't need to worry about gating such warnings behind a warning wave, because the project already won't compile.

Reply:
> We used to ignore the language version when running nullable semantic analysis. It didn't affect whether we gave diagnostics, but now it does.

I am not sure this is the behavior change that we want. It looks like this affects whether the nullability information is exposed through SemanticModel and IOperation, i.e. potentially affects behavior of analyzers and IDE features. 

---
In reply to: [462349019](https://github.com/dotnet/roslyn/pull/46402#discussion_r462349019) [](ancestors = 462349019)

Reply:
I would suggest to avoid any observable behavior changes in this PR.

---
In reply to: [462651778](https://github.com/dotnet/roslyn/pull/46402#discussion_r462651778) [](ancestors = 462651778,462349019)

Reply:
It looks like we are missing a regression test that is supposed to catch this behavior change. Consider adding one.

---
In reply to: [462654191](https://github.com/dotnet/roslyn/pull/46402#discussion_r462654191) [](ancestors = 462654191,462651778,462349019)

Reply:
I would like to add the regression test, but I was surprised that the following test passed against a1474c7. Basically it seems like `EnsureNullabilityAnalysisPerformedIfNecessary()` will catch us and build out the nullability info even if we neglected to do it during binding. I should probably sync with someone who understands the nullable semantic model better to make sure I understand the situation properly.

```cs
        [Fact]
        public void CSharp73ProvidesNullableSemanticInfo()
        {
            var source = @"
class C
{
    void M(string s)
    {
        if (s == null)
        {
            s.ToString();
        }
    }
}
";
            var comp = CreateCompilation(source, options: WithNonNullTypesTrue(), parseOptions: TestOptions.Regular7_3, skipUsesIsNullable: true);
            comp.VerifyDiagnostics(
                // error CS8630: Invalid 'NullableContextOptions' value: 'Enable' for C# 7.3. Please use language version '8.0' or greater.
                Diagnostic(ErrorCode.ERR_NullableOptionNotAvailable).WithArguments("NullableContextOptions", "Enable", "7.3", "8.0").WithLocation(1, 1)
                );

            var tree = comp.SyntaxTrees.Single();
            var model = comp.GetSemanticModel(tree);

            var node = tree.GetRoot().DescendantNodes().OfType<MemberAccessExpressionSyntax>().Single();
            var info = model.GetTypeInfo(node.Expression);
            Assert.Equal(PublicNullableAnnotation.Annotated, info.Nullability.Annotation);
            Assert.Equal(PublicNullableFlowState.MaybeNull, info.Nullability.FlowState);
        }
        
```

Reply:
> I would like to add the regression test, but I was surprised that the following test passed against a1474c7. Basically it seems like EnsureNullabilityAnalysisPerformedIfNecessary() will catch us and build out the nullability info even if we neglected to do it during binding. I should probably sync with someone who understands the nullable semantic model better to make sure I understand the situation properly.

The code in method compiler doesn't affect consumers that explicitly request semantic model from a compilation. Only those that get semantic model from an analysis context in an Analyzer action are going to be affected.

---
In reply to: [462687379](https://github.com/dotnet/roslyn/pull/46402#discussion_r462687379) [](ancestors = 462687379)